File: go/fixture/go-mini/internal/report/catalog.go
package report

import "slices"

// Device is a device known to the catalog.
type Device struct {
	Name  string
	Model string
}

// Event returns the catalog event for the device.
func (d Device) Event() Event {
	return Event{Kind: "catalog", Subject: d.Name + "/" + d.Model}
}

// Catalog is a catalog of devices.
type Catalog struct {
	devices []Device
}

// NewCatalog creates a new Catalog holding its own copy of devices.
func NewCatalog(devices []Device) *Catalog {
	return &Catalog{devices: slices.Clone(devices)}
}

// Find finds the device by name.
func (c *Catalog) Find(name string) *Device {
	for i := range c.devices {
		if c.devices[i].Name == name {
			return &c.devices[i]
		}
	}
	return nil
}

File: py/fixture/py-mini/internal/report/catalog.py
"""The devices the reporter knows how to announce."""

from dataclasses import dataclass

from internal.report.reporter import Event


@dataclass(frozen=True)
class Device:
    """A device known to the catalog."""

    name: str
    model: str

    def event(self) -> Event:
        """Return the catalog event for the device."""
        return Event(kind="catalog", subject=self.name + "/" + self.model)


class Catalog:
    """A catalog of devices."""

    def __init__(self, devices: list[Device]) -> None:
        """Create a Catalog holding its own copy of devices."""
        self._devices = list(devices)

    def find(self, name: str) -> Device | None:
        """Find the device by name."""
        for d in self._devices:
            if d.name == name:
                return d
        return None

    @classmethod
    def parse(cls, text: str) -> "Catalog":
        """Build a Catalog from catalog text, one ``name/model`` device per line."""
        devices = [d for d in map(parse_device, text.splitlines()) if d is not None]
        return cls(devices)


def parse_device(line: str) -> Device | None:
    """Parse one catalog line, written as ``name/model``.

    A blank line is not a device; it parses to None so callers can skip it.
    """
    text = line.strip()
    if not text:
        return None
    name, sep, model = text.partition("/")
    if not sep or not name or not model:
        return None
    return Device(name=name, model=model)
